contracts/governance/Badger.sol

```diff
@@ -0,0 +1,342 @@
+        string uriId;
+        bool transferable;
+    }
+
+    /*
+        State variables
+    */
+
+    mapping(uint256 => TokenTier) public tokenTiers; // tokenId => TokenTier
+
+    /*
+        Modifiers
+    */
+
+    modifier isSameLength(
+        address[] calldata accounts,
+        uint256[] calldata tokenIds,
+        uint256[] memory amounts
+    ) {
+        require(
+            accounts.length == tokenIds.length &&
+                tokenIds.length == amounts.length,
+            "Input array mismatch"
+        );
+        _;
+    }
+
+    modifier isTier(uint256 tokenId) {
+        require(
+            _isNonEmptyString(tokenTiers[tokenId].uriId),
+            "Tier does not exist"
+        );
+        _;
+    }
+
+    modifier isValidString(string calldata uriId) {
+        require(_isNonEmptyString(uriId), "URI identifier required");
+        _;
+    }
+
+    modifier isValidTransfer(uint256 tokenId, address from) {
+        require(
+            tokenTiers[tokenId].transferable,
+            "Transfer disabled for this tier"
+        );
+        require(
+            owner() == _msgSender() ||
+                from == _msgSender() ||
+                isApprovedForAll(from, _msgSender()),
+            "Unauthorized"
+        );
+        _;
+    }
+
+    /*
+        Constructor
+    */
+
+    constructor(string memory _newBaseUri) ERC1155(_newBaseUri) {}
+
+    /*
+        Minting & burning
+    */
+
+    /**
+     * @dev                 mints specified amount token(s) of specific id to specified account
+     * @param account       beneficiary address
+     * @param id            id of token, aka. tier
+     * @param amount        units of token to be minted to beneficiary
+     */
+    function mint(
+        address account,
+        uint256 id,
+        uint256 amount
+    ) public onlyOwner {
+        bytes memory data;
+
+        _mint(account, id, amount, data);
+    }
+
+    /**
+     * @dev                 burns specified amount token(s) of specific id from specified account
+     * @param account       address of token holder
+     * @param id            id of token, aka. tier
+     * @param amount        units of token to be burnt from beneficiary
+     */
+    function burn(
+        address account,
+        uint256 id,
+        uint256 amount
+    ) public onlyOwner {
+        _burn(account, id, amount);
+    }
+
+    /**
+     * @dev                 mints to multiple addresses arbitrary units of tokens of ONE token id per address
+     * @notice              example: mint 3 units of tokenId 1 to alice and 4 units of tokenId 2 to bob
+     * @param accounts      list of beneficiary addresses
+     * @param tokenIds      list of token ids (aka tiers)
+     * @param amounts       list of mint amounts
+     */
+    function mintToMultiple(
+        address[] calldata accounts,
+        uint256[] calldata tokenIds,
+        uint256[] calldata amounts
+    ) public onlyOwner isSameLength(accounts, tokenIds, amounts) {
```

Comment:
If you use modifiers, their code is basically copied over into each function that makes use of the respective modifier, they are "inlined". This increases the contract size, because on the compiled-code-level it copies the same modifier code into different places. If one would replace the modifier with an internal function, this would reduce contract size, but make each function invocation slightly more expensive, since calling an internal function is a little bit more expensive than running inlined code.

In this context a slightly larger contract is not a problem, imo.

[Here's a good article](https://blog.polymath.network/solidity-tips-and-tricks-to-save-gas-and-reduce-bytecode-size-c44580b218e6), check out the section _Function modifiers can be inefficient_